src/charm.py

```diff
@@ -0,0 +1,769 @@
+                                     secret_key: str) -> None:
+        """Update dashboard db with access & secret key for rados gateways.
+
+        This method uses the legacy format which only supports one gateway.
+        """
+        self._apply_file_setting('set-rgw-api-access-key', access_key)
+        self._apply_file_setting('set-rgw-api-secret-key', secret_key)
+
+    def _update_multi_radosgw_creds(self, creds: str) -> None:
+        """Update dashboard db with access & secret key for rados gateway."""
+        access_keys = {c['daemon_id']: c['access_key'] for c in creds}
+        secret_keys = {c['daemon_id']: c['secret_key'] for c in creds}
+        self._apply_file_setting(
+            'set-rgw-api-access-key',
+            json.dumps(access_keys))
+        self._apply_file_setting(
+            'set-rgw-api-secret-key',
+            json.dumps(secret_keys))
+
+    def _support_multiple_gateways(self) -> bool:
+        """Check if version of dashboard supports multiple rados gateways"""
+        return ch_host.cmp_pkgrevno('ceph-common', '16.0') > 0
+
+    def _manage_radosgw(self) -> None:
+        """Register rados gateways in dashboard db"""
+        if self.unit.is_leader():
+            creds = self.radosgw_user.get_user_creds()
+            cred_count = len(set([
+                (c['access_key'], c['secret_key'])
+                for c in creds]))
+            if cred_count < 1:
+                logging.info("No object gateway creds found")
+                return
+            if self._support_multiple_gateways():
+                self._update_multi_radosgw_creds(creds)
+            else:
+                if cred_count > 1:
+                    logging.error(
+                        "Cannot enable object gateway support. Ceph release "
+                        "does not support multiple object gateways in the "
+                        "dashboard")
+                else:
+                    self._update_legacy_radosgw_creds(
+                        creds[0]['access_key'],
+                        creds[0]['secret_key'])
+
+    def _request_certificates(self, event) -> None:
+        """Request TLS certificates."""
+        if not self.ca_client.is_joined:
+            logging.debug("Cannot request certificates, relation not present.")
+            return
+        addresses = set()
+        if self.ingress.relations:
+            lb_response = self.ingress.get_frontend_data()
+            if lb_response:
+                lb_config = lb_response[self.LB_SERVICE_NAME]
+                addresses.update(
+                    [i for d in lb_config.values() for i in d['ip']])
+            else:
+                logging.debug(
+                    ("Defering certificate request until loadbalancer has "
+                     "responded."))
+                event.defer()
+                return
+        for binding_name in ['public']:
+            binding = self.model.get_binding(binding_name)
```

Comment:
FYI - in public clouds like AWS, I believe this doesn't return the external IP, but rather the private IP (from within the vpc)